access-common/src/main/java/edu/unc/lib/dl/ui/service/LorisContentService.java

```diff
@@ -167,23 +176,140 @@ public void streamJP2(String simplepid, String region, String size, String rotat
+        return iiifMapper.writeValueAsString(createSequence(path, briefObjs));
+    }
+
+    public String getCanvas(HttpServletRequest request, BriefObjectMetadata briefObj)
+            throws JsonProcessingException {
+        String path = getRecordPath(request);
+        return iiifMapper.writeValueAsString(createCanvas(path, briefObj));
+    }
+
+    private Sequence createSequence(String seqPath, List<BriefObjectMetadata> briefObjs) {
+        Sequence seq = new Sequence(URIUtil.join(seqPath, "sequence", "normal"));
+
+        Set<String> uuidList = new HashSet<>();
+        for (BriefObjectMetadata briefObj : briefObjs) {
+            String datastreamUuid = jp2Pid(briefObj);
+
+            if (!datastreamUuid.equals("")) {
+                // Don't add rootObj twice
+                if (uuidList.contains(datastreamUuid)) {
+                    continue;
+                }
+
+                Canvas canvas = createCanvas(seqPath, briefObj);
+                seq.addCanvas(canvas);
+            }
+
+            uuidList.add(datastreamUuid);
+        }
+
+        return seq;
+    }
+
+    private Canvas createCanvas(String path, BriefObjectMetadata briefObj) {
+        String title = getTitle(briefObj);
+        String uuid = jp2Pid(briefObj);
+        Canvas canvas = new Canvas(path, title);
+        String canvasPath = URIUtil.join(basePath, "jp2Proxy", uuid, "jp2");
+
+        Datastream fileDs = briefObj.getDatastreamObject(DatastreamType.ORIGINAL_FILE.getId());
+        String extent = fileDs.getExtent();
+        if (extent != null) {
+            String[] imgDimensions = extent.split("x");
+            canvas.setHeight(Integer.parseInt(imgDimensions[0]));
+            canvas.setWidth(Integer.parseInt(imgDimensions[1]));
+        }
+
+        canvas.addIIIFImage(canvasPath, ImageApiProfile.LEVEL_TWO);
+        ImageContent thumb = new ImageContent(URIUtil.join(basePath,
+                "services", "api", "thumb", uuid, "large"));
+        canvas.addImage(thumb);
+
+        String label = briefObj.getLabel();
+        if (label != null) {
+            canvas.addLabel(label);
+        }
+
+        return canvas;
+    }
+
+    private String getRecordPath(HttpServletRequest request) {
         String[] url = request.getRequestURL().toString().split("\\/");
         String uuid = url[4];
         String datastream = url[5];
-        String manifestBase = URIUtil.join(basePath, uuid);
+        return URIUtil.join(basePath, "jp2Proxy", uuid, datastream);
+    }
 
-        Canvas canvas = new Canvas(manifestBase);
-        String path = URIUtil.join(basePath, "jp2Proxy", uuid, datastream);
-        canvas.addIIIFImage(path, ImageApiProfile.LEVEL_TWO);
+    private String jp2Pid(BriefObjectMetadata briefObj) {
+        Datastream datastream = briefObj.getDatastreamObject(DatastreamType.JP2_ACCESS_COPY.getId());
+        if (datastream != null) {
+            String id = datastream.getOwner();
+            if (id.equals("")) {
```

Comment:
If you return "" you miss anything that's not the primary obj. The file name, in every example I've seen, is the uuid + file extension.

Reply:
Ah okay, I understand, I was looking at the original_file datastream's filename field. Anyways, since all you need in this case is the ID of the object itself, you could return `briefObj.getId()` instead of parsing it out of the filename